### PR TITLE
Push Strains

### LINKED ISSUE #25: Push strains to Fastly
- from `deploy`: get code location
- from `yaml`: get content location
- from `yaml`: get conditions

### PR DESCRIPTION
**Pull request by trieloff**
Fixes #25 

- [x] check for errors
- [x] increase test coverage

### CODE REVIEW

File: src/deploy.js [hunk 1]
```diff
@@ -94,6 +94,11 @@ module.exports = function deploy() {
           type: 'boolean',
           default: false,
         })
+        .option('content', {
```
Comment (tripodsan) [inline src/deploy.js]: why not `contentRepo` like in helix-config.yaml ?

File: src/strain-config.js [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+const yaml = require('js-yaml');
+const hash = require('object-hash');
+
+function anon(stname) {
```
Comment (tripodsan) [inline src/strain-config.js]: anon? what does this do? jsdoc?

File: src/strain-config.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  return !!stname.match(/^[0-9a-f]{16}$/);
+}
+
+function name(strain) {
```
Comment (tripodsan) [inline src/strain-config.js]: what does this do?

File: src/strain-config.js [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+const yaml = require('js-yaml');
+const hash = require('object-hash');
+
```
Comment (tripodsan) [inline src/strain-config.js]: maybe describe what a `strain` type is here. use `@typedef` like here: https://github.com/adobe/hypermedia-pipeline/blob/master/src/pipeline.js#L24-L33

or create a `Strain` class....

File: src/strain-config.js [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+function clean(strain) {
+  if (strain.strain && strain.strain.name && !anon(strain.strain.name)) {
+    return strain.strain;
+  }
+  return { name: name(strain.strain), ...strain.strain };
+}
+
+function validate(strain) {
+  return (
+    !!strain &&
```
Comment (tripodsan) [inline src/strain-config.js]: better: `_.isObject()`

File: src/strain-config.js [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    strain.name &&
+    strain.content &&
+    strain.content.owner &&
+    strain.content.repo &&
+    strain.code
+  );
+}
+
+function wrap(strain) {
```
Comment (tripodsan) [inline src/strain-config.js]: wrap with what?

File: src/strain-config.js [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+  const anonb = anon(strainb.strain.name);
+  if (anonb && !anona) {
+    return -1;
+  } else if (anona && !anonb) {
+    return 1;
+  }
+  return straina.strain.name.localeCompare(strainb.strain.name);
+}
+
+function write(strains) {
```
Comment (tripodsan) [inline src/strain-config.js]: better: `toYaml()`; write implies writing to stream/disk/etc

File: src/strain-config.js [hunk 8]
```diff
@@ -0,0 +1,102 @@
+/*
```
Comment (tripodsan) [inline src/strain-config.js]: maybe call this `strain-config-utils`

File: src/strain.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+
+  return {
+    set executor(value) {
+      executor = value;
+    },
+    command: 'strain',
+    desc: 'Activate strains in the Fastly CDN',
+    builder: (yargs) => {
+      yargs
+        .option('wsk-auth', {
```
Comment (tripodsan) [inline src/strain.js]: duplicated code with `deploy`, I think. candiate for base class or defaults

File: test/testDeployCmd.js [hunk 10]
```diff
@@ -12,9 +12,12 @@
 
 /* eslint-env mocha */
 
+const fs = require('fs-extra');
+const assert = require('assert');
 const CLI = require('../src/cli');
 
 describe('hlx deploy (Integration)', () => {
+  fs.removeSync('.hlx');
```
Comment (tripodsan) [inline test/testDeployCmd.js]: use `before()` or `beforeEach()`

File: test/testDeployCmd.js [hunk 11]
```diff
@@ -25,6 +28,36 @@ describe('hlx deploy (Integration)', () => {
...
+        '--content', 'https://github.com/adobe/helix-cli/tree/implement-init',
+        '--target', 'test/integration/.hlx/build',
+      ]);
+    assert.ok(fs.existsSync('.hlx/strains.yaml'));
+    const thirdrun = fs.readFileSync('.hlx/strains.yaml').toString();
+    assert.notEqual(firstrun, thirdrun);
+
     done();
-  });
+  }).timeout(5000);
```
Comment (tripodsan) [inline test/testDeployCmd.js]: why 5000? what is taking so long?

File: test/testStrainConfig.js [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+- strain:
+    content:
+      owner: trieloff
+      repo: soupdemo
+      ref: master
+      root: moscow
+    code: /trieloff/default/git-github-com-adobe-helix-cli-git--dirty
+`;
+
+  const unsorted = `
```
Comment (tripodsan) [inline test/testStrainConfig.js]: why not load those strings from disk ? :-)

File: src/deploy.js (hunk 1 above)
Comment (trieloff) [reply src/deploy.js]: Because camel case command line arguments are really uncommon.

File: src/strain-config.js (hunk 8 above)
Comment (trieloff) [reply src/strain-config.js]: done

**Comment by trieloff** [comment]
@tripodsan I'm getting test failures on CircleCI and locally:

```
[hlx] info: Open soupdemo at http://localhost:61596/index.html
[hlx] debug: resolved /index.html -> html
[hlx] debug: found template at /Users/lars/Code/helix-refactor/helix-cli/test/integration/tmp/html.js
[hlx] debug: Constructing HTML Pipeline
[hlx] debug: Creating pipeline
[hlx] debug: Running HTML pipeline
[hlx] debug: fetching Markdown from http://raw.localtest.me:61595/helix/content/master/index.md
[hlx] error: Could not fetch Markdown from http://raw.localtest.me:61595/helix/content/master/index.md
```
https://circleci.com/gh/adobe/helix-cli/118 

any idea what might be causing this?

**Pull request merged**